### PR TITLE
Use Minio instead of s3 for testing

### PR DESCRIPTION
**Pull request by rozuur**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,12 +4,18 @@ python:
 before_install:
 - sudo apt-get -qq update
 - sudo apt-get install -y tree
+- mkdir -p "${HOME}"/minio/bin
```
Comment (navinpai) [inline .travis.yml]: Where is ${HOME} defined?

File: .travis.yml [hunk 2]
```diff
@@ -4,12 +4,18 @@ python:
...
+- mkdir -p "${HOME}"/minio/data
+- wget -nc -O "${HOME}"/minio/bin/minio https://dl.minio.io/server/minio/release/linux-amd64/minio
+- chmod +x "${HOME}"/minio/bin/minio
 install:
 - pip install -r requirements.txt
 script:
+# Export environment variables for s4cmd
+- export S3_SECRET_KEY=aEmMY45evpLAMMhV3GO0c2dI6PfrCqmKEeUuezVJ
```
Comment (navinpai) [inline .travis.yml]: Minio creates a access key and ID for every new server... Hardcoded keys don't work and pop up an error (https://imgur.com/a/tkxvVIy) and are anyway a bad way of handling this. Perhaps using a config file for minio (currently it creates one in ~/.minio), but we can pick from a diff location. https://docs.minio.io/docs/minio-server-configuration-guide

File: .travis.yml [hunk 3]
```diff
@@ -4,12 +4,18 @@ python:
...
+- export S4CMD_OPTS="--endpoint-url=http://127.0.0.1:9000"
+# Run minio server in background
+- "${HOME}"/minio/bin/minio server "${HOME}"/minio/data &
 - "./runtests.sh"
-env:
-  global:
-  - secure: lVuKT7TFFrWSJiPZ6gcaSmUBC8/tbZbc2L+kNEYyqT8cNyiLn6/YEE+GXVlZ98eXA7UgAqI2xd3b4j3ttERz/AfSEhJCJjcQE+wUlNZizZXudALR0Gx736sBYBC9zktU/Ngz3PnxzfgdfgbRlJRO7oP7JiQvz7DTH/bM21RKBiY=
-  - secure: ZvMrcChuDlS1u3PmkxzUnKOkkPrv28l2J8jD0wSXtd4feBFMCTTtvTinPyvOa+gyJEwn5YRM/SE3k9CvVdao5A6EHhYPSKihFPFdYB2k2jfNNDHB83YLKSHa1JiJFK8+35w2MdJDZKDQ1OAUb5jAWup0PAArqYxjNE/BCQGOCLg=
-  - secure: NjV+iU2fi4T/+UVDoKhriqqU1gowmZIyzCwB3gbbaWA/JXC2dstZlne9N2uXvN0BG5UGYM3yEhdtpkDTjUICEdxuj1/EAym45oIRT3ZJ6SaQEpYa7M+jAzcqgHWQUjSJn6hIrD0aVLRSyir8zZTa9rdZks6IJZJFjpzpCctURPw=
+- killall minio
```
Comment (navinpai) [inline .travis.yml]: Not needed since anyway the whole env is disposed off once done.

File: .travis.yml [hunk 4]
```diff
@@ -4,12 +4,18 @@ python:
 before_install:
```
Comment (navinpai) [inline .travis.yml]: Also, travis complaining about being unable to parse travis.yml. You may want to check that out. Without that, travis checks are not triggered at all.

Comment (rozuur) [reply .travis.yml]: I have fixed it now

**Comment by rozuur** [comment]
@navinpai uff, finally got it to working.
I have also raised an issue with minio about double slash test case (case14)

**Comment by navinpai** [comment]
I still see multiple test cases failing: https://travis-ci.com/bloomreach/s4cmd/builds/73840004 ... The test passes with OK but there are exceptions being thrown:

`API: GetObject(bucket=miniobucket, object=case13/source/011)
Time: 15:09:04 UTC 05/18/2018
RemoteHost: 127.0.0.1:33688
UserAgent: Boto3/1.7.19 Python/2.7.14 Linux/4.4.0-101-generic Botocore/1.10.19
Error: 'bytes=0--1' does not have a valid last byte position value
       1: cmd/object-handlers.go:135:cmd.objectAPIHandlers.GetObjectHandler()
       2: cmd/api-router.go:73:cmd.(objectAPIHandlers).GetObjectHandler-fm()
       3: /opt/go/src/net/http/server.go:1947:http.HandlerFunc.ServeHTTP()
`
Can you take a look at this?

**Comment by navinpai** [comment]
If minio differs in behaviour from s3, does it really make sense to use minio for testing? Wouldn't testing with s3 directly be more preferable (though the issue of access keys still needs to be figured out).

Minio errors are being ignored, and we have a minio specific test case as well in the test script. So even if tests pass, we have miniocmd instead of s4cmd, since CI is not testing with s3 at all. Just don't want to end up with a case where a code change works perfectly with minio, but breaks with s3, and I'm unsure how we can be sure of that.

**Comment by rozuur** [comment]
Yes, there is a chance that the code which works with minio can be broken in s3 due to incompatibilities. And one way to work around it is to execute `run_tests` with s3 before upgrading a version.

Using minio would reduce the barrier of entry for contribution and the test case which has a dependency on s3 uses double slashes in the path and the probability of it breaking is less. 

If we don't use minio we need to use companies s3 access keys and they will stop working with account termination.

**Comment by rozuur** [comment]
closed it due to incompatibilities between s3 and minio.

**Pull request closed without merging**